### ISSUE TITLE
Duo Keyboard Layout Language

### ISSUE THREAD

**Issue by ysalexg** [DuoStream/Duo#176]
Duo always seems to assign the keyboard layout to English when the user logs in, which causes me to constantly have to change it back to my language (Spanish). Is there a solution?

**Comment by nilsheidorn** [comment]
OMG, i thought i am alone & just too dumb.
Same problem here, only that my preferred anguage would be German.

Please help.
For me its just annoying but my kids kill me for it :-)

**Comment by ysalexg** [comment]
@nilsheidorn 
I solved it by running a PowerShell script at user startup. Here is the script adapted to your language; the specific changes are in the $LangList and Set-WinDefaultInputMethodOverride sections.

`$LangList = New-WinUserLanguageList de-DE
Set-WinUserLanguageList $LangList -Force

Set-WinDefaultInputMethodOverride -InputTip "0407:00000407"

$explorerProcess = Get-Process explorer

$source = @"
using System;
using System.Runtime.InteropServices;

public class KeyboardLayout
{
    [DllImport("user32.dll")]
    public static extern bool PostMessage(IntPtr hWnd, uint Msg, IntPtr wParam, IntPtr lParam);

    public static void RefreshKeyboardLayout()
    {
        const uint WM_INPUTLANGCHANGEREQUEST = 0x0050;
        PostMessage((IntPtr)0xffff, WM_INPUTLANGCHANGEREQUEST, IntPtr.Zero, IntPtr.Zero);
    }
}
"@

Add-Type -TypeDefinition $source -Language CSharp

[KeyboardLayout]::RefreshKeyboardLayout()`

**Comment by ysalexg** [comment]
@nilsheidorn 
I solved it by running a PowerShell script at user startup. Here is the script adapted to your language; the specific changes are in the $LangList and Set-WinDefaultInputMethodOverride sections.

```
$LangList = New-WinUserLanguageList de-DE
Set-WinUserLanguageList $LangList -Force

Set-WinDefaultInputMethodOverride -InputTip "0407:00000407"

$explorerProcess = Get-Process explorer

$source = @"
using System;
using System.Runtime.InteropServices;

public class KeyboardLayout
{
    [DllImport("user32.dll")]
    public static extern bool PostMessage(IntPtr hWnd, uint Msg, IntPtr wParam, IntPtr lParam);

    public static void RefreshKeyboardLayout()
    {
        const uint WM_INPUTLANGCHANGEREQUEST = 0x0050;
        PostMessage((IntPtr)0xffff, WM_INPUTLANGCHANGEREQUEST, IntPtr.Zero, IntPtr.Zero);
    }
}
"@

Add-Type -TypeDefinition $source -Language CSharp

[KeyboardLayout]::RefreshKeyboardLayout()
```

**Comment by nilsheidorn** [comment]
Thanks a lot, first thing todo tomorrow :-)

**Comment by nilsheidorn** [comment]
WOW; thank you, so friendly!

**Comment by ysalexg** [comment]
@Tecnelm 
Yes, I'm getting the same error after updating to 24H2. I have to run the script several times for it to work. I don't know if there's another solution
